File: tembu-server/tembu-server/Services/LicenseService.cs
using System.Security.Cryptography;
using System.Text;
using tembu_server.Models;

namespace tembu_server.Services;

public class LicenseService
{
    // Geheimnis — VOR Produktionseinsatz ändern, danach NIE mehr ändern
    private const string Secret = "tembu-2024-x7k9-secure";
    private const string Prefix = "TEMBU";

    public static LicenseInfo Validate(string email, string licenseKey, string expiry)
    {
        var info = new LicenseInfo();

        if (string.IsNullOrWhiteSpace(email) || string.IsNullOrWhiteSpace(licenseKey))
        {
            info.IsValid = false;
            info.ErrorMessage = "E-Mail oder Lizenzschlüssel fehlt in appsettings.json.";
            return info;
        }

        var expected = GenerateKey(email, expiry);
        if (!string.Equals(licenseKey.Trim(), expected, StringComparison.OrdinalIgnoreCase))
        {
            info.IsValid = false;
            info.ErrorMessage = "Ungültiger Lizenzschlüssel. Bitte kontaktieren Sie support@tembu.ch.";
            return info;
        }

        if (!DateTime.TryParse(expiry, out var expiryDate))
        {
            info.IsValid = false;
            info.ErrorMessage = "Ungültiges Ablaufdatum in appsettings.json (Format: YYYY-MM-DD).";
            return info;
        }

        if (DateTime.Today > expiryDate)
        {
            info.IsValid = false;
            info.ErrorMessage = $"Lizenz abgelaufen am {expiryDate:dd.MM.yyyy}. Bitte erneuern Sie Ihre Lizenz.";
            return info;
        }

        info.IsValid = true;
        info.Email = email.ToLower().Trim();
        info.ExpiryDate = expiryDate;
        return info;
    }

    public static string GenerateKey(string email, string expiry)
    {
        var input = $"{email.ToLower().Trim()}|{expiry}|{Secret}";
        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(Secret));
        var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(input));
        var hex = BitConverter.ToString(hash).Replace("-", "").ToUpper();
        return $"{Prefix}-{hex[..4]}-{hex[4..8]}-{hex[8..12]}-{hex[12..16]}";
    }
}

File: tembu-server/Services/LicenseService.vb
Imports System.Security.Cryptography
Imports System.Text

Public Class LicenseService

    ' Geheimnis — nur in Ihrem kompilierten Binary vorhanden
    ' VOR Produktionseinsatz ändern, danach NIE mehr ändern
    Private Const SECRET As String = "tembu-2024-x7k9-secure"
    Private Const PREFIX As String = "TEMBU"

    ' ── Lizenz prüfen ─────────────────────────────────────────────────────────
    Public Shared Function Validate(email As String,
                                    licenseKey As String,
                                    expiry As String,
                                    maxUsers As Integer) As LicenseInfo
        Dim info As New LicenseInfo()

        ' Eingaben prüfen
        If String.IsNullOrWhiteSpace(email) OrElse String.IsNullOrWhiteSpace(licenseKey) Then
            info.IsValid = False
            info.ErrorMessage = "E-Mail oder Lizenzschlüssel fehlt in appsettings.json."
            Return info
        End If

        ' Schlüssel prüfen
        Dim expected As String = GenerateKey(email, expiry, maxUsers)
        If Not String.Equals(licenseKey.Trim(), expected, StringComparison.OrdinalIgnoreCase) Then
            info.IsValid = False
            info.ErrorMessage = "Ungültiger Lizenzschlüssel. Bitte kontaktieren Sie support@tembu.ch."
            Return info
        End If

        ' Ablaufdatum prüfen
        Dim expiryDate As DateTime
        If Not DateTime.TryParse(expiry, expiryDate) Then
            info.IsValid = False
            info.ErrorMessage = "Ungültiges Ablaufdatum in appsettings.json (Format: YYYY-MM-DD)."
            Return info
        End If

        If DateTime.Today > expiryDate Then
            info.IsValid = False
            info.ErrorMessage = $"Lizenz abgelaufen am {expiryDate:dd.MM.yyyy}. Bitte erneuern Sie Ihre Lizenz."
            Return info
        End If

        ' Alles gültig
        info.IsValid = True
        info.Email = email.ToLower().Trim()
        info.ExpiryDate = expiryDate
        info.MaxUsers = maxUsers
        Return info
    End Function

    ' ── Schlüssel generieren (nur in Ihrem Generator-Tool verwenden) ───────────
    Public Shared Function GenerateKey(email As String,
                                       expiry As String,
                                       maxUsers As Integer) As String
        Dim input As String = $"{email.ToLower().Trim()}|{expiry}|{maxUsers}|{SECRET}"
        Using hmac As New HMACSHA256(Encoding.UTF8.GetBytes(SECRET))
            Dim hash As Byte() = hmac.ComputeHash(Encoding.UTF8.GetBytes(input))
            Dim hex As String = BitConverter.ToString(hash).Replace("-", "").ToUpper()
            Return $"{PREFIX}-{hex.Substring(0, 4)}-{hex.Substring(4, 4)}-{hex.Substring(8, 4)}-{hex.Substring(12, 4)}"
        End Using
    End Function

End Class
